packages/did-comm/src/action-handler.ts

```diff
@@ -67,8 +130,355 @@ export class DIDComm implements IAgentPlugin {
+      default:
+        throw new Error(`not_implemented: packing messages as ${args.packing} is not supported yet`)
+    }
+  }
+
+  private async packDIDCommMessageJWS(
+    args: IPackDIDCommMessageArgs,
+    context: IAgentContext<IDIDManager & IKeyManager & IResolver>,
+  ): Promise<IPackedDIDCommMessage> {
+    const message = args.message
+    let keyRef: string | undefined = args.keyRef
+    let kid: string
+    // check that the message has from field that is managed
+    let managedSender: IIdentifier | undefined
+    try {
+      managedSender = await context.agent.didManagerGet({ did: message.from || '' })
+    } catch (e) {
+      debug(`message.from(${message.from}) is not managed by this agent`)
+    }
+    if (!message.from || !isDefined(managedSender)) {
+      throw new Error('invalid_argument: `from` field must be a DID managed by this agent')
+    }
+
+    // obtain sender signing key(s) from authentication section
+    const senderKeys = await mapIdentifierKeysToDoc(managedSender, 'authentication', context)
+    // try to find a managed signing key that matches keyRef
+    let signingKey = null
+    if (isDefined(keyRef)) {
+      signingKey = senderKeys.find((key) => key.kid === keyRef || key.meta.verificationMethod.id === keyRef)
+    }
+    // otherwise use the first available one.
+    signingKey = signingKey ? signingKey : senderKeys[0]
+
+    if (!signingKey) {
+      throw new Error(`key_not_found: could not locate a suitable signing key for ${message.from}`)
+    } else {
+      kid = signingKey.meta.verificationMethod.id
+    }
+    let alg: string
+    if (signingKey.type === 'Ed25519') {
+      alg = 'EdDSA'
+    } else if (signingKey.type === 'Secp256k1') {
+      alg = 'ES256K'
+    } else {
+      throw new Error(
+        `not_supported: key of type ${signingKey.type} is not supported for JWS didcomm message`,
+      )
+    }
+    // construct the protected header with alg, typ and kid
+    const headerObj = { alg, kid, typ: DIDCommMessageMediaType.SIGNED }
+    const header = encodeJoseBlob(headerObj)
+    const payload = encodeJoseBlob(args.message)
+    // construct signing input and obtain signature
+    const signingInput = header + '.' + payload
+    const signature: string = await context.agent.keyManagerSign({
+      data: signingInput,
+      encoding: 'utf-8',
+      keyRef: signingKey.kid,
+      algorithm: alg,
+    })
+    // create flattened JWS
+    const packedMessage = {
+      protected: header,
+      payload,
+      signature,
+    }
+    // serialize flattened JWS JSON and return
+    return { message: JSON.stringify(packedMessage) }
+  }
+
+  private async packDIDCommMessageJWE(
+    args: IPackDIDCommMessageArgs,
+    context: IAgentContext<IDIDManager & IKeyManager & IResolver>,
+  ): Promise<IPackedDIDCommMessage> {
+    // 1. check if args.packing requires authentication and map sender key to skid
+    let senderECDH: ECDH | null = null
```

Comment:
Would it make sense to have some validateDIDCommMessage that validates the basic message format, e.g., `from` can be a DID URL but no `#` allowed etc. This could be used for both pack functions obviously.

Reply:
Sure, that makes sense. For the moment we rely on typescript to coerce the proper data types, but we don't do too much validation. We can add a validator.

Reply:
i think for the beta version it should be fine to just demonstrate the raw functionality. since it is beta, it is subject to change anyways.